### ISSUE TITLE
gsctl should handle multiple API endpoints as contexts

### ISSUE THREAD

**Issue by marians** [giantswarm/gsctl#47]
Story: As a user, I want to stay logged in while using more than one API endpoint, so that I can quickly switch between installations (contexts).

Currently a config file stores only one `email` and `token`. This forces users to log in every time when switching to a different API endpoint using the `--api-endpoint` flag.

We could instead store the endpoint-specific context information like this:

```yaml
endpoints:
  "https://api.giantswarm.io":
    token: ecb96eeb-3e16-4c04-8673-fb600bc36882
    email: marian@giantswarm.io
    organization: ""
    cluster: ""
updated: 2017-04-20T10:28:08+02:00
```

This would allow to easily keep login data for any number of endpoints.

**Comment by marians** [comment]
Note: With a change as described above, we'll have to introduce explicit context switching. Similar to `kubectl config set-context`, but hopefully easier to use (less typing, autocomplete, ...?)

**Comment by r7vme** [comment]
Can also be realted here: https://github.com/giantswarm/gsctl/issues/104

**Comment by marians** [comment]
Some spec as request for comments:

### Concept of endpoint switching

The "endpoint" is the API endpoint URL currently configured via the `--api-endpoint` option on every command.

Endpoint switching allows to persistently set the endpoint for all subsequent commands. The currently selected endpoint can be overridden explicitly (using a command line option).

The configuration file automatically stores each endpoint used. The token per is stored per endpoint, so that users can stay logged in with different endpoints.

As of now, gsctl assumes `https://api.giantswarm.io` as the default endpoint. That didn't prove to be useful, as most users are on installations using other endpoints, so there won't be a default endpoint any longer.

If no endpoint is configured, which will be the case on first usage, and a command is executed without `--endpoint` option, an error message will hunch the user to

- either use `gsctl login --endpoint` to add an endpoint
- or use the `--endpoint` and `--auth-token` options with the given command

### Adding an endpoint

Currently users use `gsctl login --api-endpoint ...` in order to start a session with a specific endpoint.

This will continue to work. However the `--api-endpoint` option will be deprecated. It will be replaced by `--endpoint` for consistency reasons. So the canonical way to add an endpoint will be:

```
gsctl login --endpoint https://api.mydomain.com user@mydomain.com
```

As `https://` is the only scheme we support, that port should become optional so that the above command could also be entered as `gsctl login --endpoint api.mydomain.com ...`.

The login command automatically selects the endpoint for further commands. Read on for details.

### Selecting an endpoint

One goal is to get rid of explicit endpoint setting per command using `--endpoint` (former --api-endpoint`). So users should be able to express whenever they want to switch between endpoints.

The current endpoint is selected using the new `gsctl select endpoint` command/subcommand. Example:

```
gsctl select endpoint api.giantswarm.io
```

Of course, tab-completion would be great here to simplify value entry. Not for the first iteration though.

I expect us to have other "selectable" nouns in the future. For example `gsctl select cluster ...` could be use to set a "current cluster" or `gsctl select organization ...` could be used to select the current organization. These selections could help make subsequent commands shorter, as cluster ID or organization ID could be omitted.

### Overriding the selected endpoint

To run a command against a certain endpoint, regardless of the currently selected one, the `--endpoint` option will still be available. This will not select a new endpoint. It will only set the API endpoint URL for one command execution.

### Deleting endpoints

Requires editing the `config.yaml` file.

### Getting endpoint info

The `gsctl info` command should inform about the currently selected endpoint as well as all other selectable endpoints.

## Migration

Legacy configuration files should be automatically migrated on execution of gsctl.

**Comment by webwurst** [comment]
Sounds good. I could be nice to also allow an environment variable for the endpoint. This could be set differently for each shell you are in. The global `select` will be fine for most users. But in case there is a longer running script calling `gsctl` frequently and you change the endpoint in some other shell globally via `select` this might cause harm, since the script suddenly works against a different cluster.

**Comment by xh3b4sd** [comment]
My naive two cents. IMO this should be sorted like kubectl does the context switching. The commands and the way it works should be aligned. This would make sense to me. Reinventing the wheel when there is an established way in the community does not make sense to me. Thanks for listening.

**Comment by marians** [comment]
@xh3b4sd Thanks for your comment. I have looked closely at kubectl and its "context" concept and found that it's a much more elaborate thing than what we try to accomplish. In our case, the context is really only differentiated by an endpoint URL. So my thought is we might as well just call it endpoint, as we did before, instead of introduce a new name that needs more explanation.

I hope that makes a bit of sense.

**Comment by xh3b4sd** [comment]
> In our case, the context is really only differentiated by an endpoint URL.

For now, yes. This might change faster than expected. You worked on the CLI of the version 1 platform where we had the app-service and stuff. Maybe check what we hustled with back then. IMO preparing for the future is a good thing and not really more work in this case. When I would say customers they can context switch with our tool like in kubectl this would be really cool. They would not need to relearn things, which would be a good UX IMO.

**Comment by marians** [comment]
@webwurst Thanks for the suggestion regarding the env var. I added this to the "Overriding the selected endpoint" section.

**Comment by marians** [comment]
Any more feedback here on https://github.com/giantswarm/gsctl/issues/47#issuecomment-332180593 by @cheftony @antoineco @lukaseichler @calvix @r7vme @kopiczko ?

**Comment by kopiczko** [comment]
I understand you proposed `gsctl` should keep `endpoint` to `token` mapping. I'd like it to keep `endpoint & user` to `token` mapping. Just in case we need that in future.

I don't want go into UX details, but I'd like to keep `--endpoint` or `--context` or whatever option even if `gsctl` supports global context switching. Just to be able to avoid issues @webwurst described.

**Comment by marians** [comment]
@kopiczko The simple one first:

> ... I'd like to keep --endpoint or --context or whatever option even if gsctl supports global context switching.

As explained above in section `Overriding the selected endpoint` the `--endpoint` switch will remain available. It will get higher priority than environment variable and selected endpoint.

> I understand you proposed gsctl should keep endpoint to token mapping. I'd like it to keep endpoint & user to token mapping. Just in case we need that in future.

Not sure what exactly you mean here. My intended implementation will allow you to be logged in at different endpoints at the same time. The user account used for each endpoint is independent from the rest. Different from kubectl's configuration file, with gsctl users do not maintain  configuration file with a list of user accounts and credentials. Only when logged in, we persist the auth token and the email address (per endpoint).

**Comment by kopiczko** [comment]
@marians yes, I didn't realise it wouldn't work because we don't specify user account as a parm to `gsctl`. So forget about it.

**Comment by puja108** [comment]
in general I'm ok with this, as for the topic of endpoint vs context and adding stuff in the future, we should for now already think about what happens if we add OIDC somewhere into the login mix. (and only cause it's still an alternative if at some point we might move everything behind the g8s api and actually use contexts then).

**Comment by marians** [comment]
> we should for now already think about what happens if we add OIDC somewhere into the login mix

I did in fact think about this and don't see a way how it should affect the current design.

**Comment by marians** [comment]
Update: I made up my mind regarding migrating old config files. Since the config currently doesn't contain the endpoint URL, it's not possible to migrate anything automatically.

Instead, an existing `token` and `email` key in the config should be abandoned/ignored and users will have to login before doing anything else to create a meaningful endpoint entry in the config.

**Comment by r7vme** [comment]
Looks good to me. This fixes current problem that customers had.

But for future i would ask for k8s contexts like feature. So GS engineers (and surely customers who have more that 1 installation) will be able to do like:

`gsctl context-use centaur`

**Comment by marians** [comment]
Done and released with gsctl 0.8.0
